src/game/server/neo/neo_player.cpp

```diff
@@ -3231,3 +3261,233 @@ void CNEO_Player::ModifyFireBulletsDamage(CTakeDamageInfo* dmgInfo)
 {
 	dmgInfo->SetDamage(dmgInfo->GetDamage() * sv_neo_wep_dmg_modifier.GetFloat());
 }
+
+// Start spectator takeover of player related code:
+ConVar sv_neo_spec_replace_player_bot_enable("sv_neo_spec_replace_player_bot_enable", "1", FCVAR_CHEAT, "Allow spectators to take over bots.");
+ConVar sv_neo_spec_replace_player_afk_enable("sv_neo_spec_replace_player_afk_enable", "1", FCVAR_CHEAT, "Allow spectators to take over AFK players.");
+ConVar sv_neo_spec_replace_player_afk_time_sec("sv_neo_spec_replace_player_afk_time_sec", "300", FCVAR_CHEAT, "Seconds of inactivity before a player is considered AFK.");
+
+bool CNEO_Player::IsAFK() {
+    return sv_neo_spec_replace_player_afk_enable.GetBool()
+        && (GetTimeSinceLastUserCommand() > sv_neo_spec_replace_player_afk_time_sec.GetInt());
+}
+
+void CNEO_Player::SpectatorTryReplacePlayer(CNEO_Player* pNeoPlayerToReplace)
+{
+	CSingleUserRecipientFilter filter(this);
+
+	if (!IsObserver() && IsAlive())
+	{
+		UTIL_ClientPrintFilter(filter, HUD_PRINTCONSOLE, "Spectator player takeover failed: You are not in spectator mode.");
+		return;
+	}
+
+	if (!pNeoPlayerToReplace)
+	{
+		UTIL_ClientPrintFilter(filter, HUD_PRINTCONSOLE, "Spectator player takeover failed: The spectated target is not a valid player.");
+		return;
+	}
+
+	if (!pNeoPlayerToReplace->IsAlive())
+	{
+		UTIL_ClientPrintFilter(filter, HUD_PRINTCONSOLE, "Spectator player takeover failed: The player is dead.");
+		return;
+	}
+
+	if (pNeoPlayerToReplace->GetTeamNumber() != GetTeamNumber())
+	{
+		UTIL_ClientPrintFilter(filter, HUD_PRINTCONSOLE, "Spectator player takeover failed: You must be on the same team.");
+		return;
+	}
+
+	const bool bIsTargetBot = pNeoPlayerToReplace->IsBot();
+	const bool bIsTargetAFK = pNeoPlayerToReplace->IsAFK();
+	const bool bAllowBotTakeover = sv_neo_spec_replace_player_bot_enable.GetBool();
+	const bool bAllowAfkTakeover = sv_neo_spec_replace_player_afk_enable.GetBool();
+
+	// If no valid condition is met, determine the specific reason and inform the user.
+	if (bIsTargetBot)
+	{
+		if (!bAllowBotTakeover)
+		{
+			UTIL_ClientPrintFilter(filter, HUD_PRINTCONSOLE, "Spectator player takeover failed: Taking over bots is disabled.");
+			return;
+		}
+	}
+	else if (bIsTargetAFK)
+	{
+		if (!bAllowAfkTakeover)
+		{
+			UTIL_ClientPrintFilter(filter, HUD_PRINTCONSOLE, "Spectator player takeover failed: Taking over AFK players is disabled.");
+			return;
+		}
+	}
+	else
+	{
+		UTIL_ClientPrintFilter(filter, HUD_PRINTCONSOLE, "Spectator player takeover failed: Player is neither AFK or a bot.");
+		return;
+	}
+
+	m_bSpectatorTakeoverPlayerPending = true;
+	m_hSpectatorTakeoverPlayerTarget = pNeoPlayerToReplace;
+	ForceRespawn();
+
+	// Send client confirmation to smooth out transition
+	CBroadcastRecipientFilter updatefilter;
+    updatefilter.MakeReliable();
+    UserMessageBegin(updatefilter, "CSpectatorTakeoverPlayer");
+		WRITE_SHORT(this->entindex());
+		WRITE_SHORT(pNeoPlayerToReplace->entindex());
```

Comment:
Entindex is not guaranteed unique, so we should send userid over network instead of entindex, because we have no time guarantee on when the message arrives. So technically that entindex could become invalid, or be occupied by someone else by the time the message arrives.